ksqldb-engine/src/main/java/io/confluent/ksql/schema/ksql/inference/DefaultSchemaInjector.java

```diff
@@ -91,94 +94,167 @@ public DefaultSchemaInjector(final TopicSchemaSupplier schemaSupplier) {
 
-    final SchemaResult result = schemaSupplier
-        .getValueSchema(props.getKafkaTopic(), props.getSchemaId(), expectedValueFormat);
+    if (hasValueElements(statement) || !formatSupportsSchemaInference(valueFormat)) {
+      return Optional.empty();
+    }
+
+    return Optional.of(getSchema(
+        props.getKafkaTopic(),
+        props.getValueSchemaId(),
+        valueFormat,
+        statement.getStatementText(),
+        false
+    ));
+  }
+
+  private SchemaAndId getSchema(
+      final String topicName,
+      final Optional<Integer> schemaId,
+      final FormatInfo expectedFormat,
+      final String statementText,
+      final boolean isKey
+  ) {
+    final SchemaResult result = isKey
+        ? schemaSupplier.getKeySchema(topicName, schemaId, expectedFormat)
+        : schemaSupplier.getValueSchema(topicName, schemaId, expectedFormat);
 
     if (result.failureReason.isPresent()) {
       final Exception cause = result.failureReason.get();
       throw new KsqlStatementException(
           cause.getMessage(),
-          statement.getStatementText(),
+          statementText,
           cause);
     }
 
     return result.schemaAndId.get();
   }
 
-  private static boolean hasValueElements(
+  private static boolean hasKeyElements(
       final ConfiguredStatement<CreateSource> statement
   ) {
     return statement.getStatement().getElements().stream()
-        .anyMatch(e -> e.getNamespace().equals(Namespace.VALUE));
+        .anyMatch(e -> e.getNamespace().isKey());
   }
 
-  private static boolean valueFormatSupportsSchemaInference(
+  private static boolean hasValueElements(
       final ConfiguredStatement<CreateSource> statement
   ) {
-    final FormatInfo valueFormat = SourcePropertiesUtil
-        .getValueFormat(statement.getStatement().getProperties());
+    return statement.getStatement().getElements().stream()
+        .anyMatch(e -> !e.getNamespace().isKey());
+  }
 
-    return FormatFactory.of(valueFormat).supportsFeature(SerdeFeature.SCHEMA_INFERENCE);
+  private static boolean formatSupportsSchemaInference(final FormatInfo format) {
+    return FormatFactory.of(format).supportsFeature(SerdeFeature.SCHEMA_INFERENCE);
   }
 
   private static CreateSource addSchemaFields(
       final ConfiguredStatement<CreateSource> preparedStatement,
-      final SchemaAndId schema
+      final Optional<SchemaAndId> keySchema,
+      final Optional<SchemaAndId> valueSchema
   ) {
-    final TableElements elements = buildElements(schema.columns, preparedStatement);
+    final TableElements elements = buildElements(preparedStatement, keySchema, valueSchema);
 
     final CreateSource statement = preparedStatement.getStatement();
     final CreateSourceProperties properties = statement.getProperties();
 
-    if (properties.getSchemaId().isPresent()) {
-      return statement.copyWith(elements, properties);
-    }
-    return statement.copyWith(elements, properties.withSchemaId(schema.id));
```

Comment:
This previous code had the following behavior: if `VALUE_SCHEMA_ID` was explicitly supplied by the user, then the create properties were not updated with the schema ID returned from the topic schema supplier. The new behavior is to always update the create properties with the schema ID returned from the topic schema supplier. Unless I'm mistaken, the topic schema supplier implementation always returns the same schema ID that was supplied, if relevant. Even if this changes in the future (not sure why it would), I think the new behavior is more accurate since the create properties after the injector runs reflect exactly what the engine is going to execute.